### PR TITLE
kpt fn exclude design proposal

### PR DESCRIPTION
**Pull request by natasha41575**
Design proposal for `kpt fn exclude`.

Issue: https://github.com/GoogleContainerTools/kpt/issues/2930
Implementation PR: https://github.com/GoogleContainerTools/kpt/pull/2957

### CODE REVIEW

**Review by yuwenma — APPROVED** [review]

**Review by droot — APPROVED** [review]
@natasha41575 the proposal is looking good to me. We can also get a quick review from @bgrant0607 about it.

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (bgrant0607) [inline docs/design-docs/05-kpt-fn-exclude.md:141]: Excluding local config objects would probably be a better example.

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (bgrant0607) [inline docs/design-docs/05-kpt-fn-exclude.md:165]: The SDK should make filtering out these local config objects easy.

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (bgrant0607) [inline docs/design-docs/05-kpt-fn-exclude.md:187]: I disagree this is easy to understand. It doesn't make sense to me. The right behavior entirely depends on what the function is doing. The logic really belongs in the function SDK.

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (bgrant0607) [inline docs/design-docs/05-kpt-fn-exclude.md:223]: This would break all functions that use local config pseudo-resources as implicit inputs. I had to workaround this behavior when we built the porch prototype.

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (yuwenma) [reply docs/design-docs/05-kpt-fn-exclude.md:165]: Yes, `set-namespace` excludes local-config resources inside its logic.

To have SDK filter out resources, we can check the type for Kptfile, but how can we tell which resources are function configs if not using a new annotation nor reusing local-config?

Comment (yuwenma) [reply docs/design-docs/05-kpt-fn-exclude.md:165]: To be more specific, in a hydration pipeline if function1 and function2 have different function config `funcitonConfig1` and `functionConfig2 ` correspondingly, the resourceslist for `function1` is.
```
kind: ResourceList
functionConfig: 
    kind: functionConfig1
    ...
items:
- kind: Kptfile
   ...
- kind: functionConfig2
   ...
```
the resourceslist for `function2` is.
```
kind: ResourceList
functionConfig: 
    kind: functionConfig2
    ...
items:
- kind: Kptfile
   ...
- kind: functionConfig1
   ...
```
function1 does not know which input resource is function2's config, and vice versa.

Comment (natasha41575) [reply docs/design-docs/05-kpt-fn-exclude.md:165]: In what cases will the user need to skip functionConfig? If they are our own horizontal functions like set-namespace, then setting the namespace of a functionConfig doesn't seem to really change or break any use cases. And if the user really wants to skip these functionConfigs, then they could use `kind`-based exclusion since every function has its own functionConfig kind, like `SetNamespace`. WDYT?

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (natasha41575) [reply docs/design-docs/05-kpt-fn-exclude.md]: I see your point, and totally agree that the logic belongs in the function SDK and should be controlled on a per-function level. What I was trying to say here is that the annotation is descriptive and clearly tells the user what it means. But I agree that it isn't clear or easy to understand why we would go this route, so I've taken out this line from the proposal.

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (natasha41575) [reply docs/design-docs/05-kpt-fn-exclude.md]: updated the example!

**Comment by natasha41575** [comment]
Thanks for the review @bgrant0607! I updated the proposal based on your feedback. If you have any further feedback, let me know. I plan to merge this design proposal after your approval.

File: docs/design-docs/05-kpt-fn-exclude.md
Comment (yuwenma) [reply docs/design-docs/05-kpt-fn-exclude.md:222]: My bad. I proposed this to simplify the annotations users need to deal with.

**Comment by yuwenma** [comment]
@bgrant0607  I think there's a misalignment, just want to call it out.  @natasha41575 Correct me if I myself misunderstand it.

This proposal is motived from excluding Kptfile from resources (due to recent change on including meta resource as default), but it is trying to solve a bigger problem which is described in the ["Prior art in kustomize"](https://github.com/GoogleContainerTools/kpt/pull/2971/files#diff-8b2d3c513c8efd9535c12910bfe44513d61bdd987ac5c09b2b6b4a5846cb8f84R55) that "exclude certain resources from being modified ... is one of the most popular among kustomize users". It is not limited to Kptfile, but other KRM resources users do not want the functions to touch. 

Thus, my understanding is that different customers have different resource skipping list, and it won't be a function level feature, (thus the design focus is on what's the best syntax to skip those resources easily and reliably). When you suggest this filtering to be a function level feature, do you mean the function should provide a general solution on accepting a custom FieldSpec list, iterate and skip resources from this list?  The general syntax for such a function config might be
```functionConfig.yaml
apiVersion: kpt.dev.fn/v1alpha1
kind: SetXXX
exclude:
- kind: ConfigMap
- apiVersion: kpt.dev.fn/v1alpha1
```

**Comment by natasha41575** [comment]
@yuwenma My understanding of what it means for filtering to be a "function level" feature from my discussions with @droot  is that the user should be able to control which resources are included for each function individually, rather than have kpt make the same decision for all functions in the pipeline. The design proposal currently is in alignment with that, because the user can specify `selectors`/`exclude` for each function separately.

**Comment by yuwenma** [comment]
> @yuwenma My understanding of what it means for filtering to be a "function level" feature from my discussions with @droot is that the user should be able to control which resources are included for each function individually, rather than have kpt make the same decision for all functions in the pipeline. The design proposal currently is in alignment with that, because the user can specify `selectors`/`exclude` for each function separately.

Ah, saw the other [comment](https://github.com/GoogleContainerTools/kpt/issues/2930#issuecomment-1093364459). 
I think we shall define two phases for this exclude design proposal. 
Since the current solution is more of a workaround and eventually the filtering logic should be in the SDK.  
- Phase 1. As the current proposal says, filtering the resources before passing into the function. 
- Phase 2. 
      - kpt shall annotate Kptfile and inline ConfigMap with an internal annotation, propose `internal.config.kubernetes.io/kpt-meta`, the SDK should provide functions to filter resources by annotation (and this should be called inside a function)
      - FunctionConfig defines a general interface to filter resources by GVK. SDK provides the actual filtering logic (maybe a default behavior under the SDK resourceList `Process` so it's applied to all functions automatically).  It would be like

```yaml
apiVersion: config.kubernetes.io/v1
kind: Resourcelist
items:
- apiVersion: v1
  kind: ConfigMap
  metadata:
       name: current-function-config
       annotations:
         internal.config.kubernetes.io/kpt-meta: true # marked as meta resource, exclude by default
- apiVersion: v1
  kind: ConfigMap
  metadata:
       name: other-function-config
  annotations:
       internal.config.kubernetes.io/kpt-meta: true # marked as meta resource, exclude by default
- apiVersion: kpt.dev/v1
  kind: Kptfile   # Kptfile, exclude by default
  ...
functionConfig:
  apiVersion: v1
  kind: ConfigMap
  metadata:
      name: current-function-config
      annotations:
        internal.config.kubernetes.io/exclude-meta: true
        internal.config.kubernetes.io/exclude: |     # function level exclude list.
           - kind:
           - kind:
```

**Comment by bgrant0607** [comment]
I'm fine with the exclude feature as described. It seems useful for horizontal functions like set-labels, where the function itself wouldn't necessarily know which resources should be targeted.

For the problem of changing behavior with respect to meta-resources and local resources, I see it as a temporary workaround.

Functions need to ignore resources and resource types they don't care about or would operate incorrectly on. The pipeline should always pass through everything by default.

The function SDKs need to make it straightforward to implement such filtering. 

If some resources or resource types need to be made easier to identify, I'd be fine with introducing new intra-pipeline annotations to identify them.

ConfigMaps that are generated to be used as functionConfig from inline configMaps in the Kptfile should be automatically annotated as local-config objects. ConfigMaps and pseudo-resources in their own files would need to be annotated by their creators if they didn't want them to be instantiated in the cluster.

**Comment by natasha41575** [comment]
> I'm fine with the exclude feature as described. It seems useful for horizontal functions like set-labels, where the function itself wouldn't necessarily know which resources should be targeted.

Sounds like we are all on board with the design as described, so I will go ahead and merge this PR along with the implementation PR. 

> The function SDKs need to make it straightforward to implement such filtering.

SGTM, and sounds like this would be something we discuss in Yuwen's phase 2 in [her comment](https://github.com/GoogleContainerTools/kpt/pull/2971#issuecomment-1097323671).  @yuwenma should we file a separate tracking issue for that work?

**Comment by yuwenma** [comment]
> > I'm fine with the exclude feature as described. It seems useful for horizontal functions like set-labels, where the function itself wouldn't necessarily know which resources should be targeted.
> 
> Sounds like we are all on board with the design as described, so I will go ahead and merge this PR along with the implementation PR.
> 
> > The function SDKs need to make it straightforward to implement such filtering.
> 
> SGTM, and sounds like this would be something we discuss in Yuwen's phase 2 in [her comment](https://github.com/GoogleContainerTools/kpt/pull/2971#issuecomment-1097323671). @yuwenma should we file a separate tracking issue for that work?

Filed https://github.com/GoogleContainerTools/kpt/issues/3016

**Pull request merged**
